klog.go

```diff
@@ -1638,3 +1638,21 @@ func KObjs(arg interface{}) []ObjectRef {
 	}
 	return objectRefs
 }
+
+func KObj2(obj KMetadata) ObjectRef2 {
+	return ObjectRef2{KMetadata: obj}
+}
+
+type ObjectRef2 struct {
```

Comment:
ObjectRef is expected to serialized to both text (using Stringer interface) and json (using Marshaler interface). ObjectRef annotates fields and uses default json marshaller. For ObjectRef2 we need to implement our own `MarshalJSON` method. 

Can you add `MarshalJSON` to KObj2 and also benchmark it?

Reply:
> ObjectRef is expected to serialized to both text (using Stringer interface) and json (using Marshaler interface). 

That doesn't work for json. Zap prefers the String method over MarshalJSON over marshalling structs, see my comment in https://github.com/kubernetes/klog/pull/240#issuecomment-914962514

There are no test cases with `KObj` in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/json/klog_test.go which might explain why this hasn't shown up before.

> Can you add MarshalJSON to KObj2 and also benchmark it?

I already did that in the https://github.com/pohly/klog/commits/testinglogger-kobj2 branch that I used for benchmarking, but because it never gets called, it didn't make a difference (I also ran that comparison as benchmark, just to be sure).

Reply:
Let's discuss the desired output for KObj in https://github.com/kubernetes/kubernetes/pull/104877